fix(gazelle): Do not resolve absolute imports to sibling modules

amartani opened the pull request:
Currently, gazelle allows absolute imports to be resolved to sibling modules: an `import foo` statement will resolve to a `foo.py` file in the same folder if such file exists. This seems to be a Python 2 behavior (ie. pre-`from __future__ import absolute_import`), and doesn't work on the current rules_python setup.

This behavior is explicitly tested in the [siblings_import](https://github.com/bazel-contrib/rules_python/tree/cbe6d38d01c14de46d90ea717d0f2090117533fa/gazelle/python/testdata/sibling_imports) test case. However, recreating the exact same repository layout from this test case and running `bazel test //pkg:unit_test`, the test fails with the import failing.

This PR adds a new directive, `gazelle:python_resolve_sibling_imports`, to allow disabling such behavior.

The actual changes are in 3 places:

- In `gazelle/python/target.go`, the directive is added to `if t.siblingSrcs.Contains(fileName) && fileName != filepath.Base(dep.Filepath)`, which is where the import is converted to a full absolute import if it matches a sibling file;
- In `gazelle/python/generate.go`, the handling of `conftest.py` was dependent on this behavior (ie. it added a dependency on the module `conftest`, assuming that it would be resolved to the relative module). That was modified to compute the full absolute module path instead.
- In `gazelle/python/resolve.go`, resolve relative imports even when using file generation mode.

I also explicitly added `gazelle:python_resolve_sibling_imports true` to any test that breaks if the default value of this directive is changed to `false`.

aignas commented:
Thanks for the PR. Since the current gazelle code would result in un-runnable code, is there a good reason to not default to the correct behaviour and have the flag in case some user wants to do this step-wise?

amartani commented:
> Thanks for the PR. Since the current gazelle code would result in un-runnable code, is there a good reason to not default to the correct behaviour and have the flag in case some user wants to do this step-wise?

Honestly, my main reason is a feeling of "I must be missing something" about it. The sibling import behavior is not accidental: there's code added that only exists for resolving sibling imports, a test that explicitly tests for this behavior, and multiple other tests that implicitly depend on it. And, as far as I know, it was all written much after python2 was unsupported, which is the only reason that I can think of for why this would be desired. That said, I'd be happy to switch the default if you believe that would be ok.

aignas commented:
I'm +1 for keeping the default and creating a ticket to switch the behaviour when we release 2.0.

amartani commented:
Changes since the initial PR request:
- Switched default to `python_resolve_sibling_imports false`;
- Re-worded PR title and CHANGELOG entry as a fix instead of feat / added since we are changing the default behavior;
- Adjusted behavior of relative imports when using `gazelle:experimental_allow_relative_imports true` and `gazelle:python_generation_mode file` to also resolve relative imports properly.

dougthor42 commented:
> I'm +1 for keeping the default and creating a ticket to switch the behaviour when we release 2.0.

Oooo sounds like we're about to fight!

Nah I'm kidding of course. I definitely do **not** feel strongly about adjusting the default to the correct behavior. I'm at like a 3 out of 10. So it'll be pretty easy to convince me otherwise.

aignas commented:
Ah sorry... :D I misunderstood/misread/did-not-read properly:

> Thus my vote is to adjust the default to the correct behavior: python_resolve_sibling_imports false.

+1 to go with your suggestion. Gazelle does not have the same stability guarantees as rules_python core. This breaking change should not be too disruptive the way I see it.

dougthor42 commented on gazelle/python/testdata/sibling_imports_disabled/README.md:
```diff
@@ -0,0 +1,17 @@
+- `import test_util` - resolves to the root-level `test_util.py` instead of
+  the sibling `pkg/test_util.py`
+- `from b import run` - resolves to the root-level `b.py` instead of the
+  sibling `pkg/b.py`
```
> Is there any specific case that you think it worth adding a new test for?

Nah, what is present LGTM. No action needed.

dougthor42 reviewed (approved):
Thanks, new tests LGTM.

I verified the most recent commit a00ec81df95e1a69f8d8ba2db912d946cf3f72e3 on our repo. All :+1: :tada:
